### PR TITLE
store/tikv: don't store context and don't backoff when canceled

### PR DESCRIPTION
**Pull request by tiancaiamao**
ctx should by no means been stored as member variable of a struct, 
if it's stored, it's risky that the ctx is canceled somewhere, and reused in other place.
So, ctx is removed from `RegionRequestSender` and `RPCContext`.

Backoffer is a exception, we can view it IS-A context, instead of HAS-A context, we pass it 
along call stack and `Fork()` when necessary.

Another change is that Backoff will return directly when it's canceled.

PTAL @hhkbp2 @disksing

### CODE REVIEW

**Comment by hhkbp2** [comment]
LGTM

**Comment by hhkbp2** [comment]
IMO `Backoffer` should have and only have one purpose for error retrying. It should not be IS-A context or HAS-A context. It's too much to couple `backoff` and `context` in `Backoffer` since the `context` is not related to any methods of `Backoffer`.

We could move context out of `Backoffer` later in later PR.

**Comment by hhkbp2** [comment]
PTAL @siddontang

**Comment by hhkbp2** [comment]
> I actually think that we should let backoffer become part of context, take advantage of context.WithValue().

Sounds good to me. It makes more sense than holding `context` in `Backoffer`.

**Comment by tiancaiamao** [comment]
I have a idea: we can make `Backoffer` an interface that inherit `context.Context`.
The benefit is that it's compatible with `context` while involves minimal changes.

```
type Backoffer interface {
	context.Context

	Backoff(typ backoffType, err error) error
	String() string
	Fork() (Backoffer, context.CancelFunc)
}
```

I'll do it in the next refactor.

File: store/tikv/backoff.go [hunk 1]
```diff
@@ -152,6 +152,12 @@ func NewBackoffer(maxSleep int, ctx goctx.Context) *Backoffer {
 // Backoff sleeps a while base on the backoffType and records the error message.
 // It returns a retryable error if total sleep time exceeds maxSleep.
 func (b *Backoffer) Backoff(typ backoffType, err error) error {
+	select {
+	case <-b.ctx.Done():
+		return errors.Trace(err)
```
Comment (siddontang) [inline store/tikv/backoff.go]: can the error be nil here?

Comment (tiancaiamao) [inline store/tikv/backoff.go]: if error is nil, Backoff won't be called.

**Comment by hhkbp2** [comment]
> we can make Backoffer an interface that inherit context.Context

Not sure about that. The issue involved with `Backoffer` and `context` is semantic, not grammatical. `Backoffer` should have and only have one purpose for error retrying. It should not be coupled with `context`. @tiancaiamao

**Comment by tiancaiamao** [comment]
PTAL @disksing  @siddontang

**Comment by disksing** [comment]
LGTM.

**Pull request merged**